src/query_graph/graph_path.rs

```diff
@@ -1020,6 +1154,449 @@ where
+    fn advance_with_non_collecting_and_type_preserving_transitions(
+        &self,
+        context: &OpGraphPathContext,
+        condition_resolver: &mut impl ConditionResolver,
+        excluded_destinations: &ExcludedDestinations,
+        excluded_conditions: &ExcludedConditions,
+        transition_and_context_to_trigger: impl Fn(
+            &QueryGraphEdgeTransition,
+            &OpGraphPathContext,
+        ) -> TTrigger,
+        node_and_trigger_to_edge: impl Fn(&Arc<QueryGraph>, NodeIndex, &Arc<TTrigger>) -> Option<TEdge>,
+    ) -> Result<IndirectPaths<TTrigger, TEdge>, FederationError> {
+        // If we're asked for indirect paths after an "@interfaceObject fake down cast" but that
+        // down cast comes just after non-collecting edge(s), then we can ignore the ask (skip
+        // indirect paths from there). The reason is that the presence of the non-collecting edges
+        // just before the fake down cast means we looked at indirect paths just before that
+        // down cast, but that fake down cast really does nothing in practice with the subgraph it's
+        // on, so any indirect path from that fake down cast will have a valid indirect path
+        // before it, and so will have been taken into account independently.
+        if self.last_edge_is_interface_object_fake_down_cast_after_entering_subgraph()? {
+            return Ok(IndirectPaths {
+                paths: Arc::new(vec![]),
+                dead_ends: Arc::new(Unadvanceables(vec![])),
+            });
+        }
+
+        let is_top_level_path = self.is_on_top_level_query_root()?;
+        let tail_weight = self.graph.node_weight(self.tail)?;
+        let tail_type_pos = if let QueryGraphNodeType::SchemaType(type_) = &tail_weight.type_ {
+            Some(type_)
+        } else {
+            None
+        };
+        let original_source = tail_weight.source.clone();
+        // For each source, we store the best path we find for that source with the score, or `None`
+        // if we can decide that we should not try going to that source (typically because we can
+        // prove that this create an inefficient detour for which a more direct path exists and will
+        // be found).
+        type BestPathInfo<TTrigger, TEdge> =
+            Option<(Arc<GraphPath<TTrigger, TEdge>>, QueryPlanCost)>;
+        let mut best_path_by_source: IndexMap<NodeStr, BestPathInfo<TTrigger, TEdge>> =
+            IndexMap::new();
+        let dead_ends = vec![];
+        // Note that through `excluded` we avoid taking the same edge from multiple options. But
+        // that means it's important we try the smallest paths first. That is, if we could in theory
+        // have path A -> B and A -> C -> B, and we can do B -> D, then we want to keep A -> B -> D,
+        // not A -> C -> B -> D.
+        let mut heap: BinaryHeap<HeapElement<TTrigger, TEdge>> = BinaryHeap::new();
+        while let Some(HeapElement(to_advance)) = heap.pop() {
```

Comment:
`heap` at this point is empty, so the loop never runs.

The heap should be initialised with `self`. Can we get away with storing a reference to a path in the heap? If not, you can make this function only work on `Arc<Self>` like:

```rust
fn advance_with_non_collecting_and_type_preserving_transitions(
    self: &Arc<Self>,
    ...
) {
    ...
    heap.push(Arc::clone(self));
    ...
}
```

Then users can only call this method through an `Arc<GraphPath>` and the rest of the function Just Works.

Reply:
Nice, the caller uses `Arc` so should be good!